libnavigation-base/src/main/java/com/mapbox/navigation/base/utils/DecodeUtils.kt

```diff
@@ -0,0 +1,55 @@
+@file:JvmName("DecodeUtils")
+
+package com.mapbox.navigation.base.utils
+
+import com.mapbox.api.directions.v5.DirectionsCriteria
+import com.mapbox.api.directions.v5.models.DirectionsRoute
+import com.mapbox.core.constants.Constants
+import com.mapbox.geojson.LineString
+import com.mapbox.geojson.Point
+import com.mapbox.geojson.utils.PolylineUtils
+import com.mapbox.navigation.base.internal.utils.CacheResultUtils.cacheResult
+
+private val polylineDecodeCache = PolylineUtils::decode.cacheResult(maxSize = 50)
+
+fun DirectionsRoute.completeGeometryToLinestring(): LineString? {
+    return geometry()?.geometryToLinestring(precision())
+}
+
+fun DirectionsRoute.completeGeometryToPoints(): List<Point>? {
+    return geometry()?.geometryToPoints(precision())
+}
+
+fun DirectionsRoute.stepsGeometryToLinestring(): List<List<LineString?>?>? {
+    val precision = precision()
+    return legs()?.map { leg ->
+        leg.steps()?.map { step ->
+            step.geometry()?.geometryToLinestring(precision)
+        }
+    }
+}
+
+fun DirectionsRoute.stepsGeometryToPoints(): List<List<List<Point>?>?>? {
```

Comment:
Is there a use case to return a nullable list of a nullable list of a nullable list of points?

Reply:
I don't know, I was just reflecting that list of legs, list of steps and step geometry can all be nullable. 
I can treat nulls as empty lists, if you think that'd be OK. But even in SDK itself we treat those nulls differently. 
E.g. in MapboxRouteLineUtils.parseRoutePoints we stop decoding the whole route when we encounter any null. 
And in ViewportDataSourceProcessor.processRoutePoints we treat nulls as empty lists.

Reply:
Yea that's caused by the potential `DirectionsRoute` nullability. I'd be in favor of returning an empty list - makes for fewer null-checks across the codebase.